libr/util/json.c

```diff
@@ -0,0 +1,374 @@
+						|| r_hex_to_byte (&low, p[3]) || r_hex_to_byte (&low, p[4])) {
+					R_JSON_REPORT_ERROR ("invalid unicode escape", p - 1);
+					return 0;
+				}
+				RRune codepoint = (RRune)high << 8 | (RRune)low;
+				if ((codepoint & 0xfc00) == 0xd800) { // high surrogate; need one more unicode to succeed
+					p += 6;
+					high = low = 0;
+					if (p[-1] != '\\' || *p != 'u'
+							|| r_hex_to_byte (&high, p[1]) || r_hex_to_byte (&high, p[2])
+							|| r_hex_to_byte (&low, p[3]) || r_hex_to_byte (&low, p[4])) {
+						R_JSON_REPORT_ERROR ("invalid unicode surrogate", ps);
+						return 0;
+					}
+					RRune codepoint2 = (RRune)high << 8 | (RRune)low;
+					if ((codepoint2 & 0xfc00) != 0xdc00) {
+						R_JSON_REPORT_ERROR ("invalid unicode surrogate", ps);
+						return 0;
+					}
+					codepoint = 0x10000 + ((codepoint - 0xd800) << 10) + (codepoint2 - 0xdc00);
+				}
+				int sz = r_utf8_encode ((ut8 *)d, codepoint);
+				if (!s) {
+					R_JSON_REPORT_ERROR ("invalid codepoint", ps);
+					return 0;
+				}
+				d += sz;
+				p += 5;
+				break;
+			}
+			default:
+				// leave untouched
+				*d++ = c;
+				break;
+			}
+		} else {
+			*d++ = c;
+		}
+	}
+	R_JSON_REPORT_ERROR ("no closing quote for string", s);
+	return 0;
+}
+
+static char *skip_block_comment(char *p) {
+	// assume p[-2]=='/' && p[-1]=='*'
+	char *ps = p - 2;
+	if (!*p) {
+		R_JSON_REPORT_ERROR ("endless comment", ps);
+		return 0;
+	}
+	REPEAT:
+	p = strchr (p + 1, '/');
+	if (!p) {
+		R_JSON_REPORT_ERROR ("endless comment", ps);
+		return 0;
+	}
+	if (p[-1] != '*') goto REPEAT;
+	return p + 1;
+}
+
+static char *parse_key(const char **key, char *p) {
+	// on '}' return with *p=='}'
+	char c;
+	while ((c = *p++)) {
+		if (c == '"') {
+			*key = unescape_string (p, &p);
+			if (!*key) {
+				return 0; // propagate error
+			}
+			while (*p && IS_WHITECHAR(*p)) {
+				p++;
+			}
+			if (*p == ':') {
+				return p + 1;
+			}
+			R_JSON_REPORT_ERROR ("unexpected chars", p);
+			return 0;
+		} else if (IS_WHITECHAR(c) || c == ',') {
+			// continue
+		} else if (c == '}') {
+			return p - 1;
+		} else if (c == '/') {
+			if (*p == '/') { // line comment
+				char *ps = p - 1;
+				p = strchr (p + 1, '\n');
+				if (!p) {
+					R_JSON_REPORT_ERROR ("endless comment", ps);
+					return 0; // error
+				}
+				p++;
+			} else if (*p == '*') { // block comment
+				p = skip_block_comment (p + 1);
+				if (!p) {
+					return 0;
+				}
+			} else {
+				R_JSON_REPORT_ERROR ("unexpected chars", p - 1);
+				return 0; // error
+			}
+		} else {
+			R_JSON_REPORT_ERROR ("unexpected chars", p - 1);
+			return 0; // error
+		}
+	}
+	R_JSON_REPORT_ERROR ("unexpected chars", p - 1);
+	return 0; // error
+}
+
+static char *parse_value(RJson *parent, const char *key, char *p) {
+	RJson *js;
+	while (1) {
+		switch (*p) {
+		case '\0':
+			R_JSON_REPORT_ERROR ("unexpected end of text", p);
+			return 0; // error
+		case ' ':
+		case '\t':
+		case '\n':
+		case '\r':
+		case ',':
```

Comment:
Comma is not always valid in all the contexts. This case should have extra logic to check for that

Reply:
Agreed, I have to check the best way to fix this.

Reply:
Good catch, turns out it was just ignoring all commas and treating them as whitespace, accepting stuff like `{{}{}}` or `[,,,,,,42,,,,,1337,,,,]`. I made it strict about commas now.

Reply:
Also added tests for that stuff.

Reply:
Awesome! Good work